examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_hir;
+extern crate rustc_middle;
+extern crate rustc_span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_hir::{Body, FnDecl, intravisit::FnKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self};
+use rustc_span::{
+    Span,
+    def_id::{DefId, LocalDefId},
+    symbol::Symbol,
+};
+
+dylint_linting::declare_late_lint! {
+    pub ARG_ITER,
+    Warn,
+    "functions taking `Iterator` trait bounds when `IntoIterator` would be more flexible"
+}
+
+impl<'tcx> LateLintPass<'tcx> for ArgIter {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _body: &'tcx Body<'tcx>,
+        span: Span,
+        id: LocalDefId,
+    ) {
+        // Don't lint functions generated by macros (e.g. derive, attribute macros)
+        if span.from_expansion() {
+            return;
+        }
+
+        let Some(iterator_def_id) = cx.tcx.lang_items().iterator_trait() else {
+            return; // Can't find Iterator trait definition
+        };
+
+        // Use tcx.generics_of directly
+        let generics = cx.tcx.generics_of(id);
+
+        // No generics on this function
+        if generics.own_params.is_empty() {
+            return;
+        }
+
+        let predicates = cx.tcx.predicates_of(id);
+
+        // Find generic type parameters that are *only* bound by Iterator (+ implicit Sized)
+        let mut iterator_only_params: Vec<Symbol> = Vec::new(); // Store only the name (Symbol)
+        for param in &generics.own_params {
+            // Iterate over a slice
+            if let ty::GenericParamDefKind::Type { .. } = param.kind {
+                // param.def_id is already the DefId
+                if is_param_bound_only_by_iterator(cx, predicates, param.def_id, iterator_def_id) {
+                    iterator_only_params.push(param.name); // Push only the name
+                }
+            }
+        }
+
+        if iterator_only_params.is_empty() {
+            return; // No relevant generic parameters found
+        }
+
+        // Check function inputs using rustc_middle types for robustness
+        let fn_sig = cx.tcx.fn_sig(id).skip_binder(); // Use skip_binder to handle late-bound lifetimes
+
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            // Check if the input type is a generic parameter (TyKind::Param)
+            if let ty::TyKind::Param(param_ty) = input_ty.skip_binder().kind() {
+                // Map the ParamTy index back to the GenericParamDef
+                // Use param_at since opt_param_at is deprecated and the index should be valid
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter's name is in our list of iterator-only params
+                if iterator_only_params.contains(&param_def.name) {
```

Comment:
Here, I think it should be possible to call `is_param_bound_only_by_iterator` directly, eliminating the need to call it and store its results ahead of time.

Please note that I'm recommending some changes to that function below.